tests/helper_functions.py

```diff
@@ -480,3 +481,30 @@ def __eq__(self, other):
 class AnyStringWith(str):
     def __eq__(self, other):
         return self in other
+
+
+@contextmanager
+def chdir(directory_path):
+    og_dir = os.getcwd()
+    try:
+        os.chdir(directory_path)
+        yield
+    finally:
+        os.chdir(og_dir)
+
+
+def assert_array_almost_equal(actual_array, desired_array, rtol=1e-6):
+    import numpy as np
+
+    elem0 = actual_array[0]
+    if isinstance(elem0, numbers.Number) or (
+        isinstance(elem0, (list, np.ndarray)) and isinstance(elem0[0], numbers.Number)
+    ):
+        np.testing.assert_allclose(actual_array, desired_array, rtol=rtol)
+    else:
+        np.testing.assert_array_equal(actual_array, desired_array)
+
+
+def assert_pandas_dataframe_almost_equal(actual_df, desired_df, rtol=1e-6):
```

Comment:
Can we use `pandas.testing.assert_frame_equal` with `rtol` and `atol`?

Reply:
No. It includes columns which is not number type. `assert_frame_equal` does not support it.

Reply:
Looks like `assert_frame_equal` non-numeric data types. Am I missing something?

```python
>>> from pandas.testing import assert_frame_equal
>>> import pandas as pd
>>> assert_frame_equal(pd.DataFrame({"a": ["a"]}), pd.DataFrame({"a": ["a"]}))
>>> assert_frame_equal(pd.DataFrame({"a": ["a"], "b": [0]}), pd.DataFrame({"a": ["a"], "b": [0]}))
>>> assert_frame_equal(pd.DataFrame({"a": ["a"], "b": [0]}), pd.DataFrame({"a": ["a"], "b": [1]}))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/pandas/_testing/asserters.py", line 1321, in assert_frame_equal
    check_index=False,
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/pandas/_testing/asserters.py", line 1084, in assert_series_equal
    index_values=np.asarray(left.index),
  File "pandas/_libs/testing.pyx", line 53, in pandas._libs.testing.assert_almost_equal
  File "pandas/_libs/testing.pyx", line 168, in pandas._libs.testing.assert_almost_equal
  File "/home/haru/miniconda3/envs/mlflow-dev-env/lib/python3.7/site-packages/pandas/_testing/asserters.py", line 665, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame.iloc[:, 1] (column name="b") are different

DataFrame.iloc[:, 1] (column name="b") values are different (100.0 %)
[index]: [0]
[left]:  [0]
[right]: [1]
```

Reply:
Testing `assert_frame_equal`:

```diff
diff --git a/tests/tensorflow/test_load_saved_tensorflow_estimator.py b/tests/tensorflow/test_load_saved_tensorflow_estimator.py
index 8775969dc..1317cd073 100644
--- a/tests/tensorflow/test_load_saved_tensorflow_estimator.py
+++ b/tests/tensorflow/test_load_saved_tensorflow_estimator.py
@@ -133,7 +133,7 @@ def test_iris_data_model_can_be_loaded_and_evaluated_as_pyfunc(tmpdir, model_pat
     # can call predict with a df
     results_df = pyfunc_wrapper.predict(model_data_info.inference_df)
     assert isinstance(results_df, pd.DataFrame)
-    assert_pandas_dataframe_almost_equal(results_df, model_data_info.raw_df)
+    pd.testing.assert_frame_equal(results_df, model_data_info.raw_df, check_like=True)
 
     # can also call predict with a dict
     inp_dict = {}
@@ -141,7 +141,7 @@ def test_iris_data_model_can_be_loaded_and_evaluated_as_pyfunc(tmpdir, model_pat
         inp_dict[df_col_name] = model_data_info.inference_df[df_col_name].values
     results = pyfunc_wrapper.predict(inp_dict)
     assert isinstance(results, dict)
-    assert_pandas_dataframe_almost_equal(pd.DataFrame(results), model_data_info.raw_df)
+    pd.testing.assert_frame_equal(pd.DataFrame(results), model_data_info.raw_df, check_like=True)
 
     # can not call predict with a list
     inp_list = []
```

Reply:
```
> pytest tests/tensorflow/test_load_saved_tensorflow_estimator.py -k test_iris_data_model_can_be_loaded_and_evaluated_as_pyfunc
...
================================================ 1 passed, 6 deselected, 17 warnings in 7.10s ================================================
```

Reply:
oh. I didn't know pandas.testing.assert_frame_equal supporting "rtol", let me try.
we need rtol for number columns comparison , some TF version it generates slightly different results.

Reply:
Updated. :)